spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java

```diff
@@ -202,53 +208,54 @@ public long executeTask(String taskName, Map<String, String> taskDeploymentPrope
 			}
 		}
 
+		// Build execution info and stash deploy props there
 		TaskExecutionInformation taskExecutionInformation = taskExecutionInfoService
 				.findTaskExecutionInformation(taskName, taskDeploymentProperties, composedTaskRunnerName);
-
 		if (taskExecutionInformation.isComposed()) {
 			handleAccessToken(commandLineArgs, taskExecutionInformation);
 		}
 
+		// Build app deploy request and stash deploy props there
 		TaskExecution taskExecution = taskExecutionRepositoryService.createTaskExecution(taskName);
-
 		AppDeploymentRequest appDeploymentRequest = this.taskAppDeploymentRequestCreator.
 				createRequest(taskExecution, taskExecutionInformation, commandLineArgs, platformName);
 
+		// Come up with existing and new manifests, analyze difference and update
+		// new new manifest accordingly.
 		TaskManifest taskManifest = createTaskManifest(platformName, taskExecutionInformation, appDeploymentRequest);
-
 		TaskManifest previousManifest = this.dataflowTaskExecutionMetadataDao.getLatestManifest(taskName);
 
-		if(taskDeploymentProperties.isEmpty()) {
-			if(previousManifest != null && !previousManifest.getTaskDeploymentRequest().getDeploymentProperties().equals(taskDeploymentProperties)) {
-				appDeploymentRequest = updateDeploymentProperties(commandLineArgs, platformName, taskExecutionInformation, taskExecution, previousManifest);
+		// Analysing task to know what to bring forward from existing
+		TaskAnalysisReport report = taskAnalyzer.analyze(previousManifest, taskManifest);
+		logger.debug("Task analysis report {}", report);
 
-				taskDeploymentProperties = previousManifest.getTaskDeploymentRequest().getDeploymentProperties();
-			}
-		}
+		// We now have a new props and args what should really get used.
+		Map<String, String> mergedTaskDeploymentProperties = report.getMergedDeploymentProperties();
+		List<String> mergedCommandLineArguments = report.getMergedCommandLineArguments();
+
+		taskExecutionInformation.setTaskDeploymentProperties(mergedTaskDeploymentProperties);
+		appDeploymentRequest = updateDeploymentProperties(mergedCommandLineArguments, platformName, taskExecutionInformation,
+				taskExecution, mergedTaskDeploymentProperties);
+		taskDeploymentProperties = mergedTaskDeploymentProperties;
 
 		AppDeploymentRequest request = new AppDeploymentRequest(appDeploymentRequest.getDefinition(),
```

Comment:
We can probably create the AppDeploymentRequest at this line after we updated the necessary task deployment and command line arguments.